Src/Support/Google.Apis.Auth/OAuth2/ImpersonatedCredential.cs

```diff
@@ -0,0 +1,264 @@
+    /// </para>
+    /// </summary>
+    public class ImpersonatedCredential : ServiceCredential, IOidcTokenProvider, IGoogleCredential
+    {
+        /// <summary>An initializer class for the impersonated credential. </summary>
+        new internal class Initializer : ServiceCredential.Initializer
+        {
+            /// <summary>
+            /// Gets or sets the source credential used as to acquire the impersonated credentials.
+            /// </summary>
+            public GoogleCredential SourceCredential { get; set; }
+
+            /// <summary>
+            /// Gets or sets the service account to impersonate.
+            /// </summary>
+            public string TargetPrincipal { get; set; }
+
+            /// <summary>
+            /// Gets or sets the chained list of delegates.
+            /// </summary>
+            public IEnumerable<string> Delegates { get; set; }
+
+            /// <summary>
+            /// Gets or sets the scopes to request during the authorization grant.
+            /// </summary>
+            public IEnumerable<string> Scopes { get; set; }
+
+            /// <summary>
+            /// Gets or sets the number of seconds the delegated credential should be valid.
+            /// </summary>
+            public int LifetimeInSeconds { get; set; }
+
+            /// <summary>Constructs a new initializer.</summary>
+            /// <param name="SourceCredential">The source credential used as to acquire the impersonated credentials.</param>
+            /// <param name="TargetPrincipal">The service account to impersonate.</param>
+            /// <param name="Delegates">The chained list of delegates.</param>
+            /// <param name="Scopes">The scopes to request during the authorization grant.</param>
+            /// <param name="LifetimeInSeconds">
+            /// The number of seconds the delegated credential should be valid.
+            /// By default this value should be at most 3600. However, you can follow
+            /// https://cloud.google.com/iam/docs/creating-short-lived-service-account-credentials#sa-credentials-oauth
+            /// to set up the service account and extend the maximum lifetime to 43200 (12 hours).
+            /// If the given lifetime is 0, default value 3600 will be used instead when creating the credentials.
+            /// </param>
+            public Initializer(GoogleCredential SourceCredential, string TargetPrincipal,
+                IEnumerable<string> Delegates = null, IEnumerable<string> Scopes = null, int LifetimeInSeconds = 3600)
+                : base(String.Format(GoogleAuthConsts.IamAccessTokenEndpoint, TargetPrincipal)) { 
+                    this.SourceCredential = SourceCredential;
+                    this.TargetPrincipal = TargetPrincipal;
+                    this.Delegates = Delegates;
+                    this.Scopes = Scopes;
+                    this.LifetimeInSeconds = LifetimeInSeconds;
```

Comment:
I think it would be better to do the lifetime validation here rather than in the constructor - that way we can throw an ArgumentOutOfRangeException immediately, and it'll basically be clearer.

Reply:
done